### PR TITLE
Enable reading of FileStatistics in presto-orc

### PR DESCRIPTION
**Pull request by sdruzkin**
Move intermediate encryption keys from the OrcRecordReader to the OrcReader's constructor.
Enable reading of FileStatistics for both encrypted and unencrypted ORC files.

== Test plan ==
```
mvn clean install -DskipTests
mvn clean install -pl presto-orc -pl presto-hive -pl presto-raptor
```

== RELEASE NOTES ==

General Changes
* Move IEK from OrcRecordReader to OrcReader's constructor
* Enable reading of FileStatistics for ORC files

### CODE REVIEW

**Comment by yingsu00** [comment]
> This is the reason why I had to modify the validation process in the OrcWriteValidation.
Hi @sdruzkin Will you please add reference to the PR or commit ids that you "modified the validation process in the OrcWriteValidation"?

**Review by yingsu00 — COMMENTED** [review]
@sdruzkin Sergii, I might have misunderstood your commit message that you referred to some previous commit. If you didn't mean that but just explain why you made the change in THIS commit, I think it's better to use imperative mode instead of "I had to...". 

I ran out of time reading the PR. I will continue tomorrow.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 1]
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
                 footer.hasEncryption() ? Optional.of(toEncryption(footer.getEncryption())) : Optional.empty());
     }
 
+    @Override
+    public List<ColumnStatistics> decryptFileStatistics(HiveWriterVersion hiveWriterVersion,
```
Comment (rschlussel) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:122]: This is doing more than decryption.  I would call this decryptAndCombineFileStatistics or something.

**Review by rschlussel — COMMENTED** [review]
Doing the decryption in the RecordReader is sort of ugly, though I see why you had to do it. It would be nice, though, to pass the encryption information to the OrcReader constructor instead and be able to do the decryption as part of reading the footer.  The tricky thing is that we use some info from the reader  to construct the key map(e.g the types in DwrfEncryptionMetadata.toKeyMap, and the column names in HiveUitl.getPhysicalHiveColumnHandles()).  Maybe there's a way to pass  a map that contains hive column Index + name + subfield mapped to the key into the orcreader, and then in the reader have the logic to map that to the correct Orc column.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 2]
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
+    {
+        // skip if we are not interested in any of the encrypted nodes
+        if (!dwrfEncryption.isPresent() || nodeToIntermediateKeys.isEmpty() || fileStats.isEmpty()) {
+            return fileStats;
+        }
+
+        checkState(encryptionLibrary.isPresent(), "encryption library is not set");
+        ColumnStatistics[] decryptedFileStats = fileStats.toArray(new ColumnStatistics[0]);
+
+        List<byte[]> stripeKeys = fileStripes.get(0).getKeyMetadata();
```
Comment (sujay-jain) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:139]: Is ```fileStripes``` guaranteed to be non-empty?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 3]
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
+        List<EncryptionGroup> encryptionGroups = dwrfEncryption.get().getEncryptionGroups();
+
+        // if there is a node that has child nodes then its whole subtree will be encrypted and only the parent
+        // node is added to the encryption group
+        for (int groupIdx = 0; groupIdx < encryptionGroups.size(); groupIdx++) {
+            EncryptionGroup encryptionGroup = encryptionGroups.get(groupIdx);
+            DwrfDataEncryptor decryptor = null;
+            for (int i = 0; i < encryptionGroup.getNodes().size(); i++) {
```
Comment (sujay-jain) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:147]: let's extract ```encryptionGroup.getNodes()``` to a variable so we don't call it twice

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 4]
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
+                if (!nodeToIntermediateKeys.containsKey(nodeId)) {
+                    continue;
+                }
+
+                if (decryptor == null) {
+                    // DEK for the FileStats can be stored either in the footer or/and in the first stripe.
+                    // The key in the footer takes priority over the key in the first stripe.
+                    byte[] encryptedDataKeyWithMeta;
+                    if (!encryptionGroup.getKeyMetadata().isPresent()) {
+                        encryptedDataKeyWithMeta = encryptionGroup.getKeyMetadata().get().byteArray();
```
Comment (sujay-jain) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:160]: if ```!encryptionGroup.getKeyMetadata().isPresent()``` is true then ```encryptionGroup.getKeyMetadata().get()``` will blow up?

File: presto-orc/src/test/java/com/facebook/presto/orc/TestingOrcPredicate.java [hunk 5]
```diff
@@ -83,57 +83,57 @@ public static OrcPredicate createOrcPredicate(int columnIndex, Type type, Iterab
...
+            return new BooleanOrcPredicate(columnIndex, expectedValues, false);
         }
         if (TINYINT.equals(type) || SMALLINT.equals(type) || INTEGER.equals(type) || BIGINT.equals(type)) {
             return new LongOrcPredicate(true,
                     columnIndex,
                     expectedValues.stream()
                             .map(value -> value == null ? null : ((Number) value).longValue())
                             .collect(toList()),
-                    format == DWRF);
+                    false);
```
Comment (yingsu00) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestingOrcPredicate.java:94]: `createOrcPredicate` is used by `OrcTester` and is used by all block tests. Is there any reason you set the `noFileStats` to false for all the tests? The tests could be on other formats like ORC, are they all without file stats?

**Comment by sdruzkin** [comment]
> > This is the reason why I had to modify the validation process in the OrcWriteValidation.
> 
> Hi @sdruzkin Will you please add reference to the PR or commit ids that you "modified the validation process in the OrcWriteValidation"?

Done

> @sdruzkin Sergii, I might have misunderstood your commit message that you referred to some previous commit. If you didn't mean that but just explain why you made the change in THIS commit, I think it's better to use imperative mode instead of "I had to...".

You are right, the more appropriate place for this note is the comment in the PR. I removed these lines from the commit message and PR description.

> Doing the decryption in the RecordReader is sort of ugly,
I will look at this tomorrow.

**Comment by sdruzkin** [comment]
I moved the encryption keys from the OrcRecordReader to the OrcReader in a first commit per @rschlussel's request. Then I added loading and decryption of file stats in the second commit to have clear separation between these two changes. 

I also removed changes to the validation process because now the writer validation has full access to the decrypted stats in the OrcReader.

**Comment by mbasmanova** [comment]
@sdruzkin Sergii, I see "This branch has conflicts that must be resolved". Would you rebase on top of latest master and update the PR. @rschlussel Rebecca, is this good to go otherwise?

**Review by rschlussel — APPROVED** [review]
Looks good.  Really like the new key provider abstraction.

**Pull request merged**